src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+				break;
+			case STOP:
+				ereport(LOG, (errmsg("in stop condition")));
+
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_REMOVE, &found);
+				if (found)
+				{
+					TerminateBackgroundWorker(tsbgw_he->db_scheduler_handle);
+					WaitForBackgroundWorkerShutdown(tsbgw_he->db_scheduler_handle);
+					tsbgw_total_workers_decrement();
+				}
+				tsbgw_message_send_ack(message, true);
+				break;
+
+			case RESTART:
+				ereport(LOG, (errmsg("in restart condition")));
+
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_ENTER, &found);
+				if (found)
+				{
+					TerminateBackgroundWorker(tsbgw_he->db_scheduler_handle);
+					WaitForBackgroundWorkerShutdown(tsbgw_he->db_scheduler_handle);
+				}
+				else if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+					ereport(ERROR, (errmsg("timescale background worker could not be started")));
+
+				worker_registered = register_tsbgw_entrypoint_for_db(tsbgw_he->db_oid, vxid, &tsbgw_he->db_scheduler_handle);
+				if (worker_registered)
+				{
+					pid_t		worker_pid;
+
+					WaitForBackgroundWorkerStartup(tsbgw_he->db_scheduler_handle, &worker_pid);
+					ereport(LOG, (errmsg("Worker started with PID %d", worker_pid)));
+				}
+                tsbgw_message_send_ack(message, true);
+				break;
+			default:
+				ereport(LOG, ((errmsg("timescalebgw cluster launcher unexpected message received."))));
+		}
+	}
+}
+
+extern void
+tsbgw_cluster_launcher_main(void)
+{
+
+	int         n;
+	HTAB	    *db_htab;
+
+	pqsignal(SIGTERM, tsbgw_sigterm);
+	BackgroundWorkerUnblockSignals();
+	if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+	{
+		/*
+		 * should be the first thing happening so if we already exceeded our
+		 * limits it means we have a limit of 0 and we should just exit
+		 */
+		ereport(LOG, (errmsg("Timescale Background Worker Limit Set To %d. Please set higher if you would like to use Background Workers.",TSBGW_MAX_WORKERS_GUC_STANDIN)));
+		proc_exit(0);
+	}
+	/* Connect to the db, no db name yet, so can only access shared catalogs */
+	BackgroundWorkerInitializeConnection(NULL, NULL);
+    pgstat_report_appname("Timescale BGW Cluster Launcher");
+	db_htab = populate_database_htab();
+
+    for (n = 1; n <= hash_get_num_entries(db_htab); n++ )
+        if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+        {
+            ereport(LOG, (errmsg("Total databases = %ld Timescale Background Worker Limit Set To %d. Please set higher if you would like to use Background Workers.", hash_get_num_entries(db_htab), TSBGW_MAX_WORKERS_GUC_STANDIN)));
+            proc_exit(0);
+        }
+	start_db_schedulers(db_htab);
+
+
+	while (!got_sigterm)
+	{
+		int			wl_rc;
+
+		check_for_stopped_db_schedulers(db_htab);
+		launcher_handle_messages(db_htab);
+
+		wl_rc = WaitLatch(MyLatch, WL_LATCH_SET | WL_POSTMASTER_DEATH, 0, PG_WAIT_EXTENSION);
+		ResetLatch(MyLatch);
+		if (wl_rc & WL_POSTMASTER_DEATH)
+			proc_exit(1);		/* bailout */
```

Comment:
Should we try a scheduler cleanup here before exit?

Reply:
Not on postmaster death, basically you're expected to stop everything and get the hell out if the postmaster has died, because you can end up in weird states. All workers should also bailout on postmaster death.